File: DataEngineering/03_Data_Ingestion/projects/python/03_cdc_log_tailer/main.py
import os
import json
import logging
from confluent_kafka import Consumer, KafkaException, KafkaError

logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

def create_consumer(brokers: str, group_id: str) -> Consumer:
    """Create and return a Kafka Consumer instance."""
    conf = {
        'bootstrap.servers': brokers,
        'group.id': group_id,
        'auto.offset.reset': 'earliest',
        'enable.auto.commit': False
    }
    return Consumer(conf)

def process_cdc_event(event_value: bytes):
    """Process a single CDC event."""
    try:
        payload = json.loads(event_value.decode('utf-8'))
        op = payload.get("op") # typical Debezium op code: 'c' (create), 'u' (update), 'd' (delete)
        if op in ['c', 'u']:
            logging.info(f"Event detected (INSERT/UPDATE): {payload}")
        elif op == 'd':
            logging.info(f"Event detected (DELETE): {payload}")
    except json.JSONDecodeError:
        logging.warning(f"Could not parse event as JSON: {event_value}")

def main():
    logging.info('Starting Data Engineering Project: 03_cdc_log_tailer (Kafka Consumer)')
    
    brokers = os.getenv("KAFKA_BROKERS", "localhost:9092")
    topic = os.getenv("KAFKA_CDC_TOPIC", "dbserver1.inventory.customers")
    group_id = os.getenv("KAFKA_GROUP_ID", "cdc-tailer-group")
    
    consumer = create_consumer(brokers, group_id)
    consumer.subscribe([topic])
    
    logging.info(f"Subscribed to topic {topic}. Waiting for events...")
    
    try:
        while True:
            msg = consumer.poll(timeout=1.0)
            if msg is None:
                continue
            
            if msg.error():
                if msg.error().code() == KafkaError._PARTITION_EOF:
                    # End of partition
                    continue
                else:
                    raise KafkaException(msg.error())
            
            # Process event
            process_cdc_event(msg.value())
            
            # Commit offsets manually after processing for at-least-once delivery
            consumer.commit(asynchronous=False)
            
    except KeyboardInterrupt:
        logging.info("Stopping CDC consumer...")
    except Exception as e:
        logging.error("Consumer error", exc_info=True)
    finally:
        consumer.close()

if __name__ == '__main__':
    main()

File: DataEngineering/03_Data_Ingestion/projects/golang/03_cdc_log_tailer/main.go
package main

import (
	"context"
	"fmt"
	"log"
	"os/signal"
	"syscall"
	"time"

	"github.com/segmentio/kafka-go"
)

func main() {
	log.Println("Starting Data Engineering Project: 03_cdc_log_tailer (Kafka Consumer)")

	brokers := os.Getenv("KAFKA_BROKERS")
	if brokers == "" {
		brokers = "localhost:9092"
	}

	topic := os.Getenv("KAFKA_CDC_TOPIC")
	if topic == "" {
		topic = "dbserver1.inventory.customers"
	}

	groupID := os.Getenv("KAFKA_GROUP_ID")
	if groupID == "" {
		groupID = "cdc-tailer-group"
	}

	// Setup Kafka Reader
	reader := kafka.NewReader(kafka.ReaderConfig{
		Brokers:        []string{brokers},
		GroupID:        groupID,
		Topic:          topic,
		MinBytes:       10e3, // 10KB
		MaxBytes:       10e6, // 10MB
		CommitInterval: time.Second,
		StartOffset:    kafka.FirstOffset,
	})
	defer reader.Close()

	log.Printf("Subscribed to topic %s. Waiting for events...", topic)

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	sigs := make(chan os.Signal, 1)
	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)

	go func() {
		<-sigs
		log.Println("Shutting down consumer...")
		cancel()
	}()

	for {
		m, err := reader.FetchMessage(ctx)
		if err != nil {
			if ctx.Err() != nil {
				break // Shutdown gracefully
			}
			log.Printf("Failed to fetch message: %v", err)
			time.Sleep(1 * time.Second)
			continue
		}

		fmt.Printf("CDC Event Detected at offset %d: key=%s value=%s\n", m.Offset, string(m.Key), string(m.Value))

		if err := reader.CommitMessages(ctx, m); err != nil {
			log.Printf("Failed to commit message: %v", err)
		}
	}
	log.Println("Consumer stopped cleanly.")
}
